src/nc_server.c

```diff
@@ -925,3 +932,157 @@ server_pool_deinit(struct array *server_pool)
 
     log_debug(LOG_DEBUG, "deinit %"PRIu32" pools", npool);
 }
+
+static struct msg *
+heartbeat_msg_get(struct conn *conn)
+{
+    struct msg *msg;
+
+    ASSERT(conn->client && !conn->proxy);
+
+    msg = msg_get(conn, true, conn->redis);
+    if (msg == NULL) {
+        conn->err = errno;
+    }
+
+    return msg;
+}
+
+static uint32_t
+set_heartbeat_command(struct mbuf *mbuf, int redis)
+{
+#define HEARTBEAT_MEMCACHE_COMMAND "get twemproxy\r\n"
+#define HEARTBEAT_REDIS_COMMAND "*2\r\n$3\r\nget\r\n$9\r\ntwemproxy\r\n"
+    char *command;
+    uint32_t n;
+
+    command = redis ? HEARTBEAT_REDIS_COMMAND : HEARTBEAT_MEMCACHE_COMMAND;
+    n = (uint32_t)strlen(command);
+
+    memcpy(mbuf->last, command, n);
+    ASSERT((mbuf->last + n) <= mbuf->end);
+
+    return n;
+}
+
+static rstatus_t
+send_heartbeat(struct context *ctx, struct conn *conn, struct server *server)
+{
+    struct mbuf *mbuf;
+    uint32_t n;
+    struct msg *msg;
+    struct server_pool *pool;
+    struct conn *c_conn;
+    rstatus_t status;
+
+    pool = (struct server_pool *)(server->owner);
+
+    c_conn = conn_get(pool, true, conn->redis);
+    if (c_conn == NULL) {
+        return NC_ERROR;
+    }
+
+    msg = heartbeat_msg_get(c_conn);
+    if (msg == NULL) {
+        return NC_ERROR;
+    }
+
+    c_conn->rmsg = msg;
+    mbuf = STAILQ_LAST(&msg->mhdr, mbuf, next);
+    if (mbuf == NULL || mbuf_full(mbuf)) {
+        mbuf = mbuf_get();
+        if (mbuf == NULL) {
+            return NC_ERROR;
+        }
+        mbuf_insert(&msg->mhdr, mbuf);
+        msg->pos = mbuf->pos;
+    }
+    ASSERT(mbuf->end - mbuf->last > 0);
+
+    n = set_heartbeat_command(mbuf, conn->redis);
+    mbuf->last += n;
+    msg->mlen += n;
+
+    msg->swallow = 1;
+    server->fail = FAIL_STATUS_ERR_TRY_HEARTBEAT;
+
+    if (TAILQ_EMPTY(&conn->imsg_q)) {
+        status = event_add_out(ctx->evb, conn);
+        if (status != NC_OK) {
+            return status;
+        }
+    }
+    conn->enqueue_inq(ctx, conn, msg);
+    return NC_OK;
+}
+
+void
+server_restore(struct context *ctx, struct conn *conn)
+{
+    struct server *server;
+
+    server = (struct server *)(conn->owner);
+    ASSERT(server != NULL);
+
+    if (server->fail == FAIL_STATUS_NORMAL) {
+        return;
+    }
+
+    send_heartbeat(ctx, conn, server);
```

Comment:
This will be sent both for FAIL_STATUS_ERR_TRY_HEARTBEAT and FAIL_STATUS_ERR_TRY_CONNECT - for the former, it will send redundant heartbeats even if a heartbeat is already being sent out?